src/plots/cartesian/set_convert.js

```diff
@@ -148,8 +148,13 @@ module.exports = function setConvert(ax, fullLayout) {
         // d2l/d2c variant that that won't add categories but will also
         // allow numbers to be mapped to the linearized axis positions
         var index;
-        if(ax._categoriesMap) { index = ax._categoriesMap[v] ? ax._categoriesMap : undefined; }
-        if(index !== undefined) { return index; }
+        if(ax._categoriesMap) {
+            index = ax._categoriesMap[v] ? ax._categoriesMap : undefined;
+            if(index !== undefined) return index;
+        } else {
+            index = ax._categories.indexOf(v);
+            if(index !== -1) return index;
+        }
```

Comment:
> I feel there must be some better data structure for a performant categories collection.

Good point, I'm sure there is, though it won't be built in so is unlikely to improve performance, it would just clean up the API... lets not worry about it for now, this two-part solution is light and easy enough to use here but we can look into it further if this comes up in enough other places.